### ISSUE TITLE
Run on a raspberry pi

### ISSUE THREAD

**Issue by dreamfalcon** [developersu/ns-usbloader#57]
Hi,
Tried running on a raspberry pi, but getting some errors:
sudo apt install default-jdk
sudo java -jar ns-usbloader-2.2.jar

```
OpenJDK Server VM warning: You have loaded library /root/.openjfx/cache/11/libprism_es2.so which might have disabled stack guard. The VM will try to fix the stack guard now.
It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.
Loading library prism_es2 from resource failed: java.lang.UnsatisfiedLinkError: /root/.openjfx/cache/11/libprism_es2.so: /root/.openjfx/cache/11/libprism_es2.so: wrong ELF class: ELFCLASS64 (Possible cause: architecture word width mismatch)
java.lang.UnsatisfiedLinkError: /root/.openjfx/cache/11/libprism_es2.so: /root/.openjfx/cache/11/libprism_es2.so: wrong ELF class: ELFCLASS64 (Possible cause: architecture word width mismatch)
	at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
	at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2430)
	at java.base/java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2487)
	at java.base/java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2684)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2617)
	at java.base/java.lang.Runtime.load0(Runtime.java:767)
	at java.base/java.lang.System.load(System.java:1834)
	at com.sun.glass.utils.NativeLibLoader.installLibraryFromResource(NativeLibLoader.java:205)
	at com.sun.glass.utils.NativeLibLoader.loadLibraryFromResource(NativeLibLoader.java:185)
	at com.sun.glass.utils.NativeLibLoader.loadLibraryInternal(NativeLibLoader.java:157)
	at com.sun.glass.utils.NativeLibLoader.loadLibrary(NativeLibLoader.java:52)
	at com.sun.prism.es2.ES2Pipeline.lambda$static$0(ES2Pipeline.java:68)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.sun.prism.es2.ES2Pipeline.<clinit>(ES2Pipeline.java:50)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:315)
	at com.sun.prism.GraphicsPipeline.createPipeline(GraphicsPipeline.java:187)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.init(QuantumRenderer.java:91)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:124)
	at java.base/java.lang.Thread.run(Thread.java:834)
Loading library prism_sw from resource failed: java.lang.UnsatisfiedLinkError: /root/.openjfx/cache/11/libprism_sw.so: /root/.openjfx/cache/11/libprism_sw.so: wrong ELF class: ELFCLASS64 (Possible cause: architecture word width mismatch)
java.lang.UnsatisfiedLinkError: /root/.openjfx/cache/11/libprism_sw.so: /root/.openjfx/cache/11/libprism_sw.so: wrong ELF class: ELFCLASS64 (Possible cause: architecture word width mismatch)
	at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
	at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2430)
	at java.base/java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2487)
	at java.base/java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2684)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2617)
	at java.base/java.lang.Runtime.load0(Runtime.java:767)
	at java.base/java.lang.System.load(System.java:1834)
	at com.sun.glass.utils.NativeLibLoader.installLibraryFromResource(NativeLibLoader.java:205)
	at com.sun.glass.utils.NativeLibLoader.loadLibraryFromResource(NativeLibLoader.java:185)
	at com.sun.glass.utils.NativeLibLoader.loadLibraryInternal(NativeLibLoader.java:157)
	at com.sun.glass.utils.NativeLibLoader.loadLibrary(NativeLibLoader.java:52)
	at com.sun.prism.sw.SWPipeline.lambda$static$0(SWPipeline.java:42)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.sun.prism.sw.SWPipeline.<clinit>(SWPipeline.java:41)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:315)
	at com.sun.prism.GraphicsPipeline.createPipeline(GraphicsPipeline.java:187)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.init(QuantumRenderer.java:91)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:124)
	at java.base/java.lang.Thread.run(Thread.java:834)
Graphics Device initialization failed for :  es2, sw
Error initializing QuantumRenderer: no suitable pipeline found
java.lang.RuntimeException: java.lang.RuntimeException: Error initializing QuantumRenderer: no suitable pipeline found
	at com.sun.javafx.tk.quantum.QuantumRenderer.getInstance(QuantumRenderer.java:280)
	at com.sun.javafx.tk.quantum.QuantumToolkit.init(QuantumToolkit.java:222)
	at com.sun.javafx.tk.Toolkit.getToolkit(Toolkit.java:260)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:267)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:158)
	at com.sun.javafx.application.LauncherImpl.startToolkit(LauncherImpl.java:658)
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:678)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.RuntimeException: Error initializing QuantumRenderer: no suitable pipeline found
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.init(QuantumRenderer.java:94)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:124)
	... 1 more
Exception in thread "main" java.lang.RuntimeException: No toolkit found
	at com.sun.javafx.tk.Toolkit.getToolkit(Toolkit.java:272)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:267)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:158)
	at com.sun.javafx.application.LauncherImpl.startToolkit(LauncherImpl.java:658)
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:678)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
```

Is this possible?
Thx

**Comment by developersu** [comment]
Hi,
Well. I guess it's possible. Now it's not supported. 
Anyway, good idea! I will try to add support for this =)

**Comment by dreamfalcon** [comment]
That would be great.
Thanks.

**Comment by developersu** [comment]
Hey man, 
Not much updates so far, but you may try it in CLI mode. I didn't test it yet, but it should be working. Well, except RCM-payload function, but fixing this feature won't take much time.

**Comment by dreamfalcon** [comment]
Its working, many thanks.

**Comment by msdos** [comment]
@dreamfalcon would you mind giving more information about your setup? I'm using a headless rpi, tried to run with java -jar with `sudo` but I only get a message saying NS can't be found.

Are you using a Desktop installation or headless as well?

Which nsp from switch side are you using?

Which java? I tried `openjdk-11-jre`.

(BTW, the error from the first message is because you need to install openfx, example: `sudo apt install openfx`.

**Comment by dreamfalcon** [comment]
Java version:
```
pi@raspberrypi4g:~ $ dpkg --get-selections | grep jdk
default-jdk					install
default-jdk-headless				install
openjdk-11-jdk:armhf				install
openjdk-11-jdk-headless:armhf			install
openjdk-11-jre:armhf				install
openjdk-11-jre-headless:armhf			install
```
The cmd I used:
`java -jar ns-usbloader-4.0.jar -n nsip=192.168.2.80 Mario_Tennis_Aces/hr-alera.xci`

**Comment by msdos** [comment]
Hm, I'm not using network mode, but usb.

I will try to use it in another Linux and check if it's a device problem.

**Comment by msdos** [comment]
I the end... It was a faulty cable.

I'm using Raspberry Pi OS (formerly Raspbian), headless, on Raspberry and Awoo Installer 1.3.3 on NS.

Installed

`openjdk-11-jre`

Conected the NS to the Raspberry, opened Awoo Installer and selected usb installation, and a "waiting prompt" was shown indicating that an usb connection was available to my raspberry.

Then, I ran `sudo java -jar ns-usbloader-4.0.jar -t my-nsp.nsp`. Everything worked flawlessly.

*Don't know if this is really needed*, but since it was installed in my raspberry, it may be needed (needs testing): 

`sudo apt-get install usbmount`(to automatically detect/mount the NS. If you're using a full Desktop installation this may be overkill)
`sudo at-get install openjdk-11-jdk` (didn't test without it and if only `openjdk-11-jre` would work but I'm using for other stuff)
`sudo apt install openjfx` (if you wan't to use the GUI. **I didn't test sending the nsp with the GUI**, but it may work as @dreamfalcon said. **Bonus**: even if you're headless and you want to use the GUI in a remote machine, you can configure something called "ssh X forwarding" https://unix.stackexchange.com/questions/12755/how-to-forward-x-over-ssh-to-run-graphics-applications-remotely, and it works as well).

@developersu two things. First, I believe you can close this issue. Second, please add to the docs that a faulty cable **can be** the reason for not working in some kind of troubleshooting section. I was using a usb-c cable that worked on my cell phone but **didn't work** with your software. Changing the cable made everything work.

**Comment by developersu** [comment]
Now there is a RCM payload support. I've built it on model 3B on Raspbian. Then tested. It works. Also I guess it should be working on every raspberry since they have backward-compatibility to armv6 (like aarch32). 

As for GUI.. I'm going to leave it as is. Will add a note to readme file for _**openjfx**_ dependency as a requirement for 'Raspberry Pi' GUI-mode. Thanks for testing this @dreamfalcon, I also checked this today =)

@msdos will create something like FAQ/Troubleshooting section and add this information about the cable. Thanks!
